src/diffusers/models/vq_diffusion_attention.py

```diff
@@ -0,0 +1,246 @@
+
+    Args:
+        num_embed (`int`):
+            Number of embeddings for the latent pixels embeddings.
+        height (`int`):
+            Height of the latent image i.e. the number of height embeddings.
+        width (`int`):
+            Width of the latent image i.e. the number of width embeddings.
+        embed_dim (`int`):
+            Dimension of the produced vector embeddings. Used for the latent pixel, height, and width embeddings.
+    """
+
+    def __init__(
+        self,
+        num_embed: int,
+        height: int,
+        width: int,
+        embed_dim: int,
+    ):
+        super().__init__()
+
+        self.height = height
+        self.width = width
+        self.num_embed = num_embed
+        self.embed_dim = embed_dim
+
+        self.emb = nn.Embedding(self.num_embed, embed_dim)
+        self.height_emb = nn.Embedding(self.height, embed_dim)
+        self.width_emb = nn.Embedding(self.width, embed_dim)
+
+    def forward(self, index):
+        emb = self.emb(index)
+
+        height_emb = self.height_emb(torch.arange(self.height, device=index.device).view(1, self.height))
+
+        # 1 x H x D -> 1 x H x 1 x D
+        height_emb = height_emb.unsqueeze(2)
+
+        width_emb = self.width_emb(torch.arange(self.width, device=index.device).view(1, self.width))
+
+        # 1 x W x D -> 1 x 1 x W x D
+        width_emb = width_emb.unsqueeze(1)
+
+        pos_emb = height_emb + width_emb
+
+        # 1 x H x W x D -> 1 x L xD
+        pos_emb = pos_emb.view(1, self.height * self.width, -1)
+
+        emb = emb + pos_emb[:, : emb.shape[1], :]
+
+        return emb
+
+
+class VQDiffusionTransformerBlock(nn.Module):
+    """
+    A basic transformer block modified to take the timestep of the diffusion process as additional input.
+    """
+
+    def __init__(
+        self,
+        dim: int,
+        n_heads: int,
+        d_head: int,
+        context_dim: int,
+        diffusion_steps: int,
+        dropout=0.0,
+    ):
+        super().__init__()
+
+        self.attn1 = CrossAttention(
+            query_dim=dim, heads=n_heads, dim_head=d_head, dropout=dropout, bias=True
+        )  # is a self-attention
+        self.ff = VQDiffusionFeedForward(dim=dim, dropout=dropout)
+        self.attn2 = CrossAttention(
+            query_dim=dim, context_dim=context_dim, heads=n_heads, dim_head=d_head, dropout=dropout, bias=True
+        )
+        self.norm1 = AdaLayerNorm(dim, diffusion_steps)
+        self.norm2 = AdaLayerNorm(dim, diffusion_steps)
+        self.norm3 = nn.LayerNorm(dim)
+
+    def forward(self, hidden_states, context, timestep):
+        hidden_states = self.attn1(self.norm1(hidden_states, timestep)) + hidden_states
+        hidden_states = self.attn2(self.norm2(hidden_states, timestep), context=context) + hidden_states
+        hidden_states = self.ff(self.norm3(hidden_states)) + hidden_states
+        return hidden_states
+
+
+class VQDiffusionFeedForward(nn.Module):
```

Comment:
That's quite a normal feed-forward class I think - could we maybe try to re-use: https://github.com/huggingface/diffusers/blob/888468dd90b4e353225d7d505fa28d2963e65678/src/diffusers/models/attention.py#L345 (adding some more default values that are set here